File: src/types/magicalActions/ElvenMagicalSong.ts
import * as DB from "tsondb/schema/dsl"
import { OldParameter } from "../_ActivatableSkill.js"
import { DurationUnitValue } from "../_ActivatableSkillDuration.js"
import { ActivatableSkillEffect } from "../_ActivatableSkillEffect.js"
import { PropertyIdentifier, SkillIdentifier } from "../_Identifier.js"
import { ImprovementCost } from "../_ImprovementCost.js"
import { ResponsiveText, ResponsiveTextReplace } from "../_ResponsiveText.js"
import { SkillCheck, SkillCheckPenalty } from "../_SkillCheck.js"
import { NestedTranslationMap } from "../Locale.js"
import { Errata } from "../source/_Erratum.js"
import { src } from "../source/_PublicationRef.js"

export const ElvenMagicalSong = DB.Entity(import.meta.url, {
  name: "ElvenMagicalSong",
  namePlural: "ElvenMagicalSongs",
  type: () =>
    DB.Object({
      check: DB.Required({
        comment: "Lists the linked three attributes used to make a skill check.",
        type: DB.IncludeIdentifier(SkillCheck),
      }),
      check_penalty: DB.Optional({
        comment: "In some cases, the target's Spirit or Toughness is applied as a penalty.",
        type: DB.IncludeIdentifier(SkillCheckPenalty),
      }),
      parameters: DB.Required({
        comment: "Measurable parameters of an Elven magical song.",
        type: DB.IncludeIdentifier(ElvenMagicalSongPerformanceParameters),
      }),
      skill: DB.Required({
        comment:
          "To enhance their songs, elves can make a check on either *Singing (Two-Voiced Singing)* or *Music (appropriate application)* (or both) before making the check for the song.",
        type: DB.Array(SkillIdentifier(), {
          minItems: 1,
          maxItems: 2,
          uniqueItems: true,
        }),
      }),
      property: DB.Required({
        comment: "The associated property.",
        type: PropertyIdentifier(),
      }),
      improvement_cost: DB.Required({
        comment: "States which column is used to improve the skill.",
        type: DB.IncludeIdentifier(ImprovementCost),
      }),
      src,
      translations: NestedTranslationMap(
        DB.Required,
        "ElvenMagicalSong",
        DB.Object({
          name: DB.Required({
            comment: "The Elven magical song’s name.",
            type: DB.String({ minLength: 1 }),
          }),
          effect: DB.Required({
            comment:
              "The effect description may be either a plain text or a text that is divided by a list of effects for each quality level. It may also be a list for each two quality levels.",
            type: DB.IncludeIdentifier(ActivatableSkillEffect),
          }),
          cost: DB.Optional({
            isDeprecated: true,
            type: DB.IncludeIdentifier(OldParameter),
          }),
          errata: DB.Optional({
            type: DB.IncludeIdentifier(Errata),
          }),
        }),
      ),
    }),
  instanceDisplayName: {},
  uniqueConstraints: [
    {
      entityMapKeyPath: "translations",
      keyPathInEntityMap: "name",
    },
  ],
})

const ElvenMagicalSongPerformanceParameters = DB.TypeAlias(import.meta.url, {
  name: "ElvenMagicalSongPerformanceParameters",
  comment: "Measurable parameters of an Elven magical song.",
  type: () =>
    DB.Object({
      cost: DB.Required({
        comment: "The AE cost.",
        type: DB.IncludeIdentifier(ElvenMagicalSongCost),
      }),
    }),
})

const ElvenMagicalSongCost = DB.TypeAlias(import.meta.url, {
  name: "ElvenMagicalSongCost",
  type: () =>
    DB.Object({
      value: DB.Required({
        comment: "The (temporary) AE cost value.",
        type: DB.Integer({ minimum: 1 }),
      }),
      interval: DB.Optional({
        comment: "Specified if the AE cost `value` has to be paid for each time interval.",
        type: DB.IncludeIdentifier(DurationUnitValue),
      }),
      permanent: DB.Optional({
        comment: "A permanent AE cost, independent from a possible interval.",
        type: DB.IncludeIdentifier(ElvenMagicalSongPermanentCost),
      }),
      translations: NestedTranslationMap(
        DB.Optional,
        "ElvenMagicalSongCost",
        DB.Object({
          per: DB.Required({
            comment: "The cost have to be per a specific countable entity, e.g. `8 AE per person`.",
            type: DB.IncludeIdentifier(ResponsiveText),
          }),
        }),
      ),
    }),
})

const ElvenMagicalSongPermanentCost = DB.TypeAlias(import.meta.url, {
  name: "ElvenMagicalSongPermanentCost",
  type: () =>
    DB.Object({
      value: DB.Required({
        comment: "The permanent AE cost value.",
        type: DB.Integer({ minimum: 1 }),
      }),
      translations: NestedTranslationMap(
        DB.Optional,
        "ElvenMagicalSongPermanentCost",
        DB.Object({
          replacement: DB.Required({
            comment: "A replacement string for the permanent cost.",
            type: DB.IncludeIdentifier(ResponsiveTextReplace),
          }),
        }),
      ),
    }),
})

File: Sources/OptolithDatabaseSchema/Types/MagicalActions/ElvenMagicalSong.swift
import FileDB

@Model
public struct ElvenMagicalSong {

  /// Lists the linked three attributes used to make a skill check.
  let check: SkillCheck

  /// In some cases, the target's Spirit or Toughness is applied as a penalty.
  let check_penalty: SkillCheckPenalty?

  /// Measurable parameters of an Elven magical song.
  let parameters: ElvenMagicalSongPerformanceParameters
      /// To enhance their songs, elves can make a check on either *Singing (Two-Voiced Singing)* or *Music (appropriate application)* (or both) before making the check for the song.
      @MinItems(1)
      @MaxItems(2)
      @UniqueItems
      @Relationship(Skill.self)
      let skill: [Skill.ID]

  /// The associated property.
  @Relationship(Property.self)
  let property: Property.ID

  /// States which column is used to improve the skill.
  let improvement_cost: ImprovementCost

    /// The publications where you can find the entry.
    @MinItems(1)
    let src: [PublicationRef]

    /// All translations for the entry, identified by IETF language tag (BCP47).
    @Relationship(Locale.self)
    let translations: [String: Translation]

    @Embedded
    struct Translation { // ElvenMagicalSongTranslation

        /// The Elven magical song’s name.
        @MinLength(1)
        let name: String
          /// The effect description may be either a plain text or a text that is divided by a list of effects for each quality level. It may also be a list for each two quality levels.
          let effect: ActivatableSkillEffect

        @available(*, deprecated, message: "Use language-independent performance parameters instead")
        let cost: OldParameter

        /// A list of errata for the entry in the specific language.
        @MinItems(1)
        let errata: [Erratum]?
    }
}

/// Measurable parameters of an Elven magical song.
@Embedded
public struct ElvenMagicalSongPerformanceParameters {

  /// The AE cost.
  let cost: ElvenMagicalSongCost
  }

@Embedded
public struct ElvenMagicalSongCost {

  /// The (temporary) AE cost value.
  @Minimum(1)
  let value: Int

  /// Specified if the AE cost `value` has to be paid for each time interval.
  let interval: DurationUnitValue

  /// A permanent AE cost, independent from a possible interval.
  let permanent: ElvenMagicalSongPermanentCost?
      /// All translations for the entry, identified by IETF language tag (BCP47).
      @Relationship(Locale.self)
      let translations: [String: Translation]?

      struct Translation { // ElvenMagicalSongCostTranslation

        /// The cost have to be per a specific countable entity, e.g. `8 AE per person`.
        let per: ResponsiveText
      }
  }

@Embedded
public struct ElvenMagicalSongPermanentCost {

  /// The permanent AE cost value.
  @Minimum(1)
  let value: Int
      /// All translations for the entry, identified by IETF language tag (BCP47).
      @Relationship(Locale.self)
      let translations: [String: Translation]?

      struct Translation { // ElvenMagicalSongPermanentCostTranslation

        /// A replacement string for the permanent cost.
        let replacement: ResponsiveTextReplace
      }
  }
